sdk/eventhubs/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessorAsyncClient.java

```diff
@@ -0,0 +1,140 @@
+                value.close();
+                partitionConsumers.remove(key);
+            } catch (IOException ex) {
+                logger.warning("Unable to close event hub consumer for partition {}", key);
+            }
+        });
+        disposable.dispose();
+        scheduler.dispose();
+        logger.info("Stopping the event processor");
+    }
+
+    /* Internal implementation. This is only a simple demo to show how it may look
+    - will need to move it to a different class and add more details to this */
+    void run() {
+        /* This should run periodically get new ownership details and close/open new
+        consumers when ownership of this instance has changed */
+
+        List<String> eventHubPartitionIds = new ArrayList<>();
+        this.eventHubAsyncClient.getPartitionIds().doOnNext(eventHubPartitionIds::add).blockLast();
+        logger.info("Found {} partitions", eventHubPartitionIds.size());
+
+        // Get the current ownership information from underlying data store
+        List<PartitionOwnership> ownership = new ArrayList<>();
+        this.partitionManager
+            .listOwnership("", this.consumerGroupName)
+            .doOnNext(ownership::add)
+            .blockLast();
+
+        logger.info("Got ownership info from partition manager");
+        List<PartitionOwnership> myOwnership = new ArrayList<>();
+
+        // Find the partitions to claim ownership
+        this.partitionManager
+            .claimOwnership(ownership)
+            .doOnNext(myOwnership::add)
+            .blockLast();
+        logger.info("Claimed ownership");
+
+        // use the successfully owned partition list to create new consumers
+        // and receive events
+
+        // For poc, create consumer for all partitions and start from earliest event position.
+        // For actual implementation, consumers will be created for newly owned partitions only and will
+        // start from the sequence number coming from the data store or from the initial event position (if none exists)
+        eventHubPartitionIds.forEach(this::receiveEvents);
+    }
+
+    /**
+     * Creates a new consumer for given partition and starts receiving events for that partition
+     */
+    private void receiveEvents(String partitionId) {
+        EventHubConsumer consumer = this.eventHubAsyncClient
+            .createConsumer(this.consumerGroupName, partitionId, this.initialEventPosition);
+        this.partitionConsumers.put(partitionId, consumer);
+        PartitionContext partitionContext = new PartitionContext()
+            .partitionId(partitionId)
+            .eventHubName("")
+            .consumerGroupName(this.consumerGroupName);
+        CheckpointManager checkpointManager = new CheckpointManager(partitionContext, this.partitionManager, null);
+
+        logger.info("Subscribing to receive events from partition {}", partitionId);
+        consumer.receive().subscribe(
```

Comment:
.subscribe() returns a subscription, should we be keeping track of these and .disposing them when the program ends?
ie. with `Disposables.composite()`